### DISCUSSION TITLE
DNS requests failing with 5s or 10s timeouts

### DISCUSSION THREAD

**Discussion by colinmollenhour** [moby/moby#47660] [Q&A]
A discussion was started on #46199 that was tangentially related [here](https://github.com/moby/moby/issues/46199#issuecomment-2032338440) so I'm opening a discussion per request from @akerouanton.

Specifically, I am experiencing 5 and 10 second response times for "dig", particularly when using Quad9 DNS resolvers. As suggested I ran `conntrack -S` and the results looked like this:

```
cpu=0           found=199 invalid=6283 ignore=471766 insert=0 insert_failed=0 drop=0 early_drop=0 error=1 search_restart=1345222
cpu=1           found=215 invalid=5843 ignore=475113 insert=0 insert_failed=0 drop=0 early_drop=0 error=2 search_restart=1336034
cpu=2           found=193 invalid=5533 ignore=460339 insert=0 insert_failed=0 drop=0 early_drop=0 error=0 search_restart=1354211
cpu=3           found=163 invalid=8573 ignore=304773 insert=0 insert_failed=0 drop=0 early_drop=0 error=1 search_restart=1262725
cpu=4           found=188 invalid=7160 ignore=392107 insert=0 insert_failed=0 drop=0 early_drop=0 error=1 search_restart=1361169
cpu=5           found=199 invalid=6210 ignore=394692 insert=0 insert_failed=0 drop=0 early_drop=0 error=0 search_restart=1326756
cpu=6           found=234 invalid=20245 ignore=297731 insert=0 insert_failed=0 drop=0 early_drop=0 error=1 search_restart=1758928
cpu=7           found=317 invalid=890722 ignore=1022973 insert=0 insert_failed=2662 drop=2592 early_drop=0 error=9270 search_restart=8901692
cpu=8           found=227 invalid=38666 ignore=241491 insert=0 insert_failed=0 drop=0 early_drop=0 error=4 search_restart=2462636
cpu=9           found=161 invalid=872341 ignore=919601 insert=0 insert_failed=2818 drop=2731 early_drop=0 error=10146 search_restart=8731573
cpu=10          found=143 invalid=45082 ignore=221160 insert=0 insert_failed=0 drop=0 early_drop=0 error=4 search_restart=2487178
cpu=11          found=152 invalid=899143 ignore=907548 insert=0 insert_failed=3409 drop=3295 early_drop=0 error=9345 search_restart=8939111
cpu=12          found=163 invalid=34762 ignore=359433 insert=0 insert_failed=0 drop=0 early_drop=0 error=1 search_restart=2186805
cpu=13          found=183 invalid=36447 ignore=469003 insert=0 insert_failed=0 drop=0 early_drop=0 error=0 search_restart=2339250
cpu=14          found=180 invalid=31410 ignore=444650 insert=0 insert_failed=0 drop=0 early_drop=0 error=0 search_restart=2152859
cpu=15          found=158 invalid=8235 ignore=299441 insert=0 insert_failed=0 drop=0 early_drop=0 error=0 search_restart=1245792
cpu=16          found=196 invalid=7225 ignore=387385 insert=0 insert_failed=0 drop=0 early_drop=0 error=0 search_restart=1375222
cpu=17          found=200 invalid=6873 ignore=380749 insert=0 insert_failed=0 drop=0 early_drop=0 error=0 search_restart=1349991
...
```

To which Albin responded:

> Those insert_failed could mean there're clashes over conntrack entry creation (as described in https://blog.quentin-machu.fr/2018/06/24/5-15s-dns-lookups-on-kubernetes/).

I don't understand the issue at all at this point but am happy to continue discussion and provide info and run tests as needed.

Thanks!

**Comment by akerouanton** [comment]
@colinmollenhour Could you copy the output of `docker info` please? I'm expecting more fields in the output of `conntrack -S`, especially `clashres` (but the matching kernel field was added in v5.10, so maybe yours is too old). I'll dig kernel source code according to your kernel version to make sure I interpret these fields correctly.

Also, could you try to run `conntrack -S` a first time, then run your test script to reproduce the issue, re-run `conntrack -S` and finally diff both outputs to see which fields get updated. It'd be great if you could paste the output of your test script and of that `conntrack -S` diff.

**Reply by colinmollenhour** [reply]
In the case above it is running a quite-old version of Docker.. 19.03.13. However, I also reproduced the issue on a newer server running 25.0.3 and actually the issue can be produced on the host without Docker at all so I didn't mean to imply this was an issue with Docker. The original statement I was trying to make is that in general DNS is not 100% reliable so if it supported a mechanism to serve stale records that would be a great improvement.

I can't at the moment but will try to get the diff as requested soon.
